File: app/src/main/java/org/monksanctum/xand11/extension/ExtensionProtocol.java
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.extension;

import org.monksanctum.xand11.Client;
import org.monksanctum.xand11.Dispatcher;
import org.monksanctum.xand11.comm.PacketReader;
import org.monksanctum.xand11.comm.PacketWriter;
import org.monksanctum.xand11.comm.Request;
import org.monksanctum.xand11.comm.XProtoWriter.WriteException;

import java.util.Set;

public class ExtensionProtocol implements Dispatcher.PacketHandler {

    private static final byte[] HANDLED_OPS = new byte[] {
            Request.QUERY_EXTENSION,
            Request.LIST_EXTENSIONS,
    };

    private ExtensionManager mExtensionManager;

    public ExtensionProtocol(ExtensionManager manager) {
        mExtensionManager = manager;
    }

    @Override
    public byte[] getOpCodes() {
        return HANDLED_OPS;
    }

    @Override
    public void handleRequest(Client client, PacketReader reader, PacketWriter writer) {
        switch (reader.getMajorOpCode()) {
            case Request.QUERY_EXTENSION:
                handleQueryExtension(reader, writer);
                break;
            case Request.LIST_EXTENSIONS:
                handleListExtension(reader, writer);
                break;
        }
    }

    private void handleQueryExtension(PacketReader packet, PacketWriter writer) {
        int n = packet.readCard16();
        packet.readPadding(2);
        String name = packet.readPaddedString(n);
        ExtensionManager.ExtensionInfo info = mExtensionManager.queryExtension(name);
        try {
            info.write(writer);
        } catch (WriteException e) {
        }
    }

    private void handleListExtension(PacketReader packet, PacketWriter writer) {
        int count = 0;
        Set<String> exts = mExtensionManager.getExtensions();
        writer.setMinorOpCode((byte) exts.size());
        writer.writePadding(24);
        for (String name : exts) {
            writer.writeString(name);
            writer.writeByte((byte) 0);
            count += name.length() + 1;
        }
        while ((count++ % 4) != 0) {
            writer.writePadding(1);
        }
    }
}

File: core/src/main/java/com/monksanctum/xand11/core/extension/ExtensionProtocol.kt
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.extension

import org.monksanctum.xand11.comm.*
import org.monksanctum.xand11.comm.XProtoWriter.WriteException

class ExtensionProtocol(private val mExtensionManager: ExtensionManager) : Dispatcher.PacketHandler {

    override val opCodes: ByteArray
        get() = HANDLED_OPS

    override fun handleRequest(client: Client, reader: PacketReader, writer: PacketWriter) {
        when (reader.majorOpCode) {
            Request.QUERY_EXTENSION.code -> handleQueryExtension(reader, writer)
            Request.LIST_EXTENSIONS.code -> handleListExtension(reader, writer)
        }
    }

    private fun handleQueryExtension(packet: PacketReader, writer: PacketWriter) {
        val n = packet.readCard16()
        packet.readPadding(2)
        val name = packet.readPaddedString(n)
        val info = mExtensionManager.queryExtension(name)
        try {
            info.write(writer)
        } catch (e: WriteException) {
        }

    }

    private fun handleListExtension(packet: PacketReader, writer: PacketWriter) {
        var count = 0
        val exts = mExtensionManager.extensions
        writer.minorOpCode = exts.size.toByte()
        writer.writePadding(24)
        for (name in exts) {
            writer.writeString(name)
            writer.writeByte(0.toByte())
            count += name.length + 1
        }
        while (count++ % 4 != 0) {
            writer.writePadding(1)
        }
    }

    companion object {

        private val HANDLED_OPS = byteArrayOf(Request.QUERY_EXTENSION.code,
                Request.LIST_EXTENSIONS.code)
    }
}
